Remote runtime configuration with DAPR concept (#49)
kyma-project/control-plane: docs/internal/investigations/dapr-remote-runtime-configuration/configuration_concept.md

> # Passing configuration to runtimes concept

Author: What do you think about `Remote runtime management concept`?

## Overview

This document describes the concept of passing a configuration to runtimes from the central point with the usage of [DAPR](https://dapr.io).

> ![Concept image](assets/concept.png?raw=true "Concept")

Author: Well, I thought of this document as a description where the concept of Runtime management lays in the big picture of Kyma environment. I'd go with more specific diagrams on the POC doc. What do you thing about that?

## Idea

> The main idea is to have a component in Kyma Control Plane (Runtime Director on image), which would be a point of communication with Runtimes through Agents and provide a configuration for them. The configuration would be for example an URL and credentials to some external service (e.g. logging service). After each runtime fetch the new configuration, the DAPR bindings should be updated and populated to the DAPR sidecars.

Reviewer: I have an impression from that sentence that the main reason of having the Runtime Director component is to have a possibility to deliver configuration to the runtimes. I would rephrase this to more underline the intention of having the Runtime Director for runtime management. Then in some details we can focus on the configuration part. 

I'm also missing the big picture describing the architecture presented on the concept image.

> ## Example

Reviewer: This example fits more to the actual POC documentation, describing the actual use-case to implement.

# Remote runtime management concept

## Overview

This document describes the concept of runtime management with the usage of [Dapr](https://dapr.io).

## Architecture

![Concept image](assets/concept.png?raw=true "Concept")

The diagram above shows the future architecture of Kyma environment.
There are two central places, Compass and Kyma Control Plane along with runtimes which will communicate with them.

### Compass
It is responsible for connecting applications to runtimes, along with controlling and monitoring them.

> The System Broker will be responsible for fetching data from different types of runtimes (e.g. Cloud Foundry or customer Kubernetes cluster with service catalog installed on it).  

Author: Good idea, I will change that sentence

### Kyma Control Plane

> It is responsible for provisioning new Kyma runtimes and providing initial configurations for them. 

Reviewer: ```suggestion
The Kyma Control Plane is responsible for managing the Kyma instances. It allows to provision Kyma Runtimes as well as it is a central place where the Kyma instances can fetch their configuration from. 
```

> The Runtime Director will be a service that will be responsible for managing and controlling runtimes through runtime agents.

Reviewer: ```suggestion
The Runtime Director exposes the API for the Kyma Runtimes configuration. 
```

### Runtimes

> The diagram presents some possible runtimes, which will be supported in the future. It includes Cloud Foundry runtime, Customer own cluster with Service Catalog installed, Kyma Managed Runtime, and Kyma Standalone Runtime. 

Reviewer: ```suggestion
The diagram shows possible runtimes, which will be supported. It includes Cloud Foundry runtime, Customer own cluster with Service Catalog installed, Kyma Managed Runtime, and Kyma Standalone Runtime. 
```

> Each of them beside Kyma Standalone will have some implementation of Agent, that will communicate with Control Plane.

Reviewer: ```suggestion
Each supported Kyma instance has a dedicated agent component that is responsible for handling the communication with the Kyma Central Plane. The agent component fetches the configuration from the Runtime Director and applies it to the desired environment.
```

> ## Kyma Runtime configuration from the Control Plane

Author: Actually this is true, except for the Dapr part. It is not said that every runtime would be using Dapr even if we decided to go with it for Managed Kyma Runtimes. I will rephrase this sentence

> The main idea is to have a component in Kyma Control Plane (Runtime Director on image), which would be a point of communication with Runtimes through Agents to manage them. It would allow also to configure them from one central point.

Reviewer 2: ```suggestion
The main idea is to have a component in Kyma Control Plane that will be a point of communication for Agents to manage the Runtimes. This component would be the image of Runtime Director. It would also allow for configuring Runtimes from one central place.
```

> The configuration could be for example an URL and credentials to some external service (e.g. logging service) which is used by some of the registered Runtimes. After each runtime fetches the new configuration, the Dapr bindings would be updated and populated to the Dapr sidecars.

Reviewer 2: ```suggestion
For example, the configuration could be passed as an URL with credentials to some external service, such as logging service which is used by some of the registered Runtimes. After each Runtime fetches the new configuration, the Dapr bindings would be updated and populated to the Dapr sidecars.
```

# Remote Runtime management concept

> 

Reviewer 2: ```suggestion
```

### Kyma Control Plane
Kyma Control Plane is responsible for managing Kyma instances. It allows you to provision Kyma Runtimes and it's also a central place where Kyma instances can fetch their configuration from. 
The Runtime Director component exposes the API for the Kyma Runtimes configuration. 

### Runtimes
As shown in the diagram, these Runtimes will be supported:
- Cloud Foundry 
- customer own Kubernetes cluster with Service Catalog installed
- managed Kyma Runtime
- standalone Kyma Runtime
Each supported Kyma instance has a dedicated Agent component that is responsible for handling the communication with the Kyma Central Plane. The Agent component fetches the configuration from the Runtime Director and applies it to the desired environment.

## Managing Runtimes configuration in Kyma Control Plane
The main idea is to have a component in Kyma Control Plane that will be a point of communication for Agents to manage the Runtimes. This component would be the image of Runtime Director. It would also allow for configuring Runtimes from one central place.

> For example, the configuration could be passed as an URL with credentials to some external service, such as logging service which is used by some of the registered Runtimes. In case of Managed Kyma Runtimes, after each Runtime fetches the new configuration, the Dapr bindings would be updated and populated to the Dapr sidecars.

Reviewer 2: ```suggestion
For example, the configuration could be passed as an URL with credentials to some external service, such as logging service which is used by some of the registered Runtimes. In the case of managed Kyma Runtimes, after each Runtime fetches the new configuration, the Dapr bindings would be updated and populated to the Dapr sidecars.
```
